### PR TITLE
Changes from my comment and more

### PR DESCRIPTION
**Pull request by leudz**
Using `collapse_these` in map operations added an unnecessary `Copy` bound to all closures. I phrased it before by being "too restrictive" because any closure not `Copy` (and `FnOnce` usually aren't) can't be used. `fold_these` had this problem twice too.

I renamed `this_option` and similar based on `Result`'s methods name, [Result::ok](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok) and [Result::err](https://doc.rust-lang.org/std/result/enum.Result.html#method.err). In general, accessors in Rust are just named what they access.

I modified `is_this` and similar to take a reference instead of taking ownership. For non `Copy` types it would have consumed `These` to return the boolean.

I added a lot of pattern matching, it's so powerful, why deprive ourselves? Look at `These::here` for example. I think it makes methods easier to understand individually.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -565,9 +562,13 @@ impl<T, U> These<T, U> {
     /// ```
     ///
     /// [`This`]: enum.These.html#variant.This
-    pub fn is_this(self) -> bool
+    pub fn is_this(&self) -> bool
     {
-        self.this_option().is_some()
+        if let These::This(_) = self {
```
Comment (FintanH) [inline src/lib.rs:567]: Ah ya, this is relatively new syntax to me! I need to keep it in mind in the future :)

**Review by FintanH — APPROVED** [review]
Thanks! 

> Using collapse_these in map operations added an unnecessary Copy bound to all closures. I phrased it before by being "too restrictive" because any closure not Copy (and FnOnce usually aren't) can't be used. fold_these had this problem twice too.

Ya so this was really me playing with the idea of writing higher-order functions and seeing the limitations. This feedback is what I needed. I wasn't sure what the implications were for adding the `Copy` trait there. Do you have any resources where I could see what happens when you add `Copy` to functions?

> I renamed this_option and similar based on Result's methods name, Result::ok and Result::err. In general, accessors in Rust are just named what they access.

Makes sense, I'll stick with the idiomatic way so 👌 

> I modified is_this and similar to take a reference instead of taking ownership. For non Copy types it would have consumed These to return the boolean.

Ya, I thought about this and wasn't happy with it. Glad you showed me a clean way of writing it!

> I added a lot of pattern matching, it's so powerful, why deprive ourselves? Look at These::here for example. I think it makes methods easier to understand individually.

Don't get me wrong, I love pattern matching and ADTs. But I also love abstracting away patterns. In this case the pattern is pattern matching to apply a function, which `collapse_these` handles. Due to the way Rust passes around functions though it seems like it's not the right way to do things; which is fine :)

Again thanks a lot! This has been a serious help for me in understanding Rust a bit more 😄

**Comment by leudz** [comment]
[This article](https://stevedonovan.github.io/rustifications/2018/08/18/rust-closures-are-hard.html) explains how closures are implemented. [The reference](https://doc.rust-lang.org/reference/types/closure.html#other-traits) lists all traits implemented by closures and in which circumstance.

Also if don't know it, [clippy](https://github.com/rust-lang/rust-clippy) is a great tool. It will indicate `if let` instead of `match` for single pattern for example.

Glad I could help, feel free to ask **any** question on reddit (you already know this one 😄) or on [the forum](https://users.rust-lang.org/), there is always people happy to answer.

**Comment by FintanH** [comment]
Thanks for the links @leudz ❤️

**Pull request merged**